internal/controller/gitsync_controller.go

```diff
@@ -39,19 +73,190 @@ type GitSyncReconciler struct {
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.16.3/pkg/reconcile
 func (r *GitSyncReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = logging.FromContext(ctx)
+	logger := logging.FromContext(ctx)
+
+	// since this function can be called by multiple goroutines at the same time, lock so we only process a given GitSync one at a time
+	lockAsInterface, foundLock := r.gitSyncLocks.Load(req.NamespacedName.String())
+	if !foundLock {
+		r.gitSyncLocks.Store(req.NamespacedName.String(), &sync.Mutex{})
+		lockAsInterface, _ = r.gitSyncLocks.Load(req.NamespacedName.String())
+	}
+	lock := lockAsInterface.(*sync.Mutex)
+	lock.Lock()
+	defer lock.Unlock()
+
+	// get the GitSync CRD - if not found, it may have been deleted in the past
+	gitSync := &apiv1.GitSync{}
+	if err := r.Client.Get(ctx, req.NamespacedName, gitSync); err != nil {
+		// if we aren't able to do a Get, then either it's been deleted in the past, or something else went wrong
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		} else {
+			logger.Errorw("Unable to get GitSync", "err", err, "request", req)
+			return ctrl.Result{}, err
+		}
+	}
+
+	gitSyncOrig := gitSync // save this off so we can compare it later
+	gitSync = gitSync.DeepCopy()
+
+	gitSync.Status.InitConditions()
+
+	result, err := r.reconcile(ctx, gitSync)
+	if err != nil {
+		return result, err
+	}
+
+	// Update the Spec if needed
+	if needsUpdate(gitSyncOrig, gitSync) {
+		gitSyncStatus := gitSync.Status
+		if err := r.Client.Update(ctx, gitSync); err != nil {
+			logger.Errorw("Error Updating GitSync", "err", err, "GitSync", gitSync)
+			return ctrl.Result{}, err
+		}
+		// restore the original status, which would've been wiped in the previous call to Update()
+		gitSync.Status = gitSyncStatus
+
+	}
+
+	// Update the Status subresource
+	if gitSync.DeletionTimestamp.IsZero() { // would've already been deleted
+		if err := r.Client.Status().Update(ctx, gitSync); err != nil {
+			logger.Errorw("Error Updating GitSync Status", "err", err, "GitSync", gitSync)
+			return ctrl.Result{}, err
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// reconcile does the real logic
+func (r *GitSyncReconciler) reconcile(ctx context.Context, gitSync *apiv1.GitSync) (ctrl.Result, error) {
+	logger := logging.FromContext(ctx)
+
+	// We will have a GitSyncProcessor object for every GitSync that contains our cluster
+	// Therefore, we Create or Update it if our cluster is defined as a Destination in the GitSync
+	// We Delete it if either the GitSync is deleted or our cluster is removed from it
+
+	forThisCluster := gitSync.DeletionTimestamp.IsZero() && gitSync.Spec.ContainsClusterDestination(r.clusterName)
+	_, forThisClusterPrev := r.gitSyncProcessors.Load(gitSync.String())
+	shouldDelete := !forThisCluster && forThisClusterPrev
+	shouldAdd := forThisCluster && !forThisClusterPrev
+	shouldUpdate := forThisCluster && forThisClusterPrev
```

Comment:
I think this check here also needs to contain namespace comparison?

Reply:
How so? If we're listed as a cluster, we want a GitSyncProcessor set up. If we're not listed as a cluster, we don't. If the namespace were to change, that's something that the `GitSyncProcessor` will handle (making sure that the resources get transferred to the new namespace). Or at least that was my thought.

Reply:
yeah, after a second thought, you're right. This is the checking for actions on GitSync CRs. And we only want GitSync CR to be cluster level thus namespace option should be ignored.  This actually leads to another question, that if somehow there are two GitSync CR with the same git repo path and destination in two separate namespaces. We should not create two GitSyncProcessors, right? Even one of them cannot apply the resources successfully. In that case we probably want to maintain a cluster level mapping between GitSync and GitSyncProcessor instead of namespaced one at L48.

Reply:
So, just to make sure we're on the same page, line 48 is basically maintaining the map of the namespaced name of the GitSync CRDs themselves. At this point, for our MVP our parent-level GitSync will indicate a single namespace (maybe "numaplane-system" itself), so all of the GitSyncs will basically end up in there. 

Here we're looking at the destinations described in the GitSync CRD itself, so these will be in other namespaces besides "numaplane-system" in general.

Reply:
So, you asked about if there are two GitSync CRs with the same git repo path and destination in two separate namespaces? Do you still have this question after my clarification?

Reply:
Actually, one thing I'm realizing I should do is have a good distinction in the Status for whether the CRD is applicable/not-applicable for this cluster.

Reply:
> Actually, one thing I'm realizing I should do is have a good distinction in the Status for whether the CR is applicable/not-applicable for this cluster.

Yeah, that sounds good.

Reply:
Thanks @juliev0 for the explanation! As we discussed offline. The thing I'm worried about is two GitSync CR unnecessarily watching the same git repo path and we definitely don't need to address it here. I will create an issue for it.